chain/chain/src/resharding/manager.rs

```diff
@@ -84,34 +82,102 @@ impl ReshardingManager {
             tracing::debug!(target: "resharding", ?next_shard_layout, "next shard layout is not v2, skipping");
             return Ok(());
         }
+
         let resharding_event_type =
             ReshardingEventType::from_shard_layout(&next_shard_layout, *block_hash, *prev_hash)?;
-        let Some(ReshardingEventType::SplitShard(split_shard_event)) = resharding_event_type else {
-            tracing::debug!(target: "resharding", ?resharding_event_type, "resharding event type is not split shard, skipping");
-            return Ok(());
+        match resharding_event_type {
+            Some(ReshardingEventType::SplitShard(split_shard_event)) => {
+                self.split_shard(
+                    chain_store_update,
+                    block,
+                    shard_uid,
+                    tries,
+                    split_shard_event,
+                    next_shard_layout,
+                )?;
+            }
+            None => {
+                tracing::warn!(target: "resharding", ?resharding_event_type, "unsupported resharding event type, skipping");
+            }
         };
+        Ok(())
+    }
+
+    fn split_shard(
+        &mut self,
+        chain_store_update: ChainStoreUpdate,
+        block: &Block,
+        shard_uid: ShardUId,
+        tries: ShardTries,
+        split_shard_event: ReshardingSplitShardParams,
+        next_shard_layout: ShardLayout,
+    ) -> Result<(), Error> {
         if split_shard_event.parent_shard != shard_uid {
             let parent_shard = split_shard_event.parent_shard;
-            tracing::debug!(target: "resharding", ?parent_shard, "shard uid does not match event parent shard, skipping");
+            tracing::debug!(target: "resharding", ?parent_shard, "ShardUId does not match event parent shard, skipping");
             return Ok(());
         }
 
-        // TODO(resharding): what if node doesn't have memtrie? just pause
-        // processing?
-        // TODO(resharding): fork handling. if epoch is finalized on different
-        // blocks, the second finalization will crash.
-        tries.freeze_mem_tries(
+        // Reshard the State column by setting ShardUId mapping from children to parent.
+        self.set_state_shard_uid_mapping(&split_shard_event)?;
+
+        // Create temporary children memtries by freezing parent memtrie and referencing it.
+        self.process_memtrie_resharding_storage_update(
+            chain_store_update,
+            block,
             shard_uid,
-            vec![split_shard_event.left_child_shard, split_shard_event.right_child_shard],
+            tries,
+            split_shard_event.clone(),
         )?;
 
         // Trigger resharding of flat storage.
         self.flat_storage_resharder.start_resharding(
-            ReshardingEventType::SplitShard(split_shard_event.clone()),
+            ReshardingEventType::SplitShard(split_shard_event),
             &next_shard_layout,
         )?;
```

Comment:
I really like how this is shaping up - split_shard consists of splitting the State, FlatState and MemTrie.